core/src/main/java/hudson/diagnosis/ReverseProxySetupMonitor.java

```diff
@@ -62,26 +67,35 @@ public boolean isActivated() {
     }
 
     public HttpResponse doTest() {
-        String referer = Stapler.getCurrentRequest().getReferer();
         Jenkins j = Jenkins.getInstance();
+        String referer = Stapler.getCurrentRequest().getReferer();
         // May need to send an absolute URL, since handling of HttpRedirect with a relative URL does not currently honor X-Forwarded-Proto/Port at all.
-        String redirect = j.getRootUrl() + "administrativeMonitor/" + id + "/testForReverseProxySetup/" + (referer != null ? Util.rawEncode(referer) : "NO-REFERER") + "/";
+        
+        String redirect = buildRedirect(referer, j);
```

Comment:
(the `~~` are used as a delimiter because they're non-reserved special characters that won't appear in a referrer URL and cause extraction to fail)